### DISCUSSION TITLE
Unable to run ansible-playbook -c local azure-provision.yaml

### DISCUSSION THREAD

**Discussion by harryadwani** [Sunbird-Ed/Community#31] [Installation] [answered]
Trying to install on Azure and using free tier. This is the error I get on cloud shell. I am completely new to sunbird.

`TASK [Create virtual network] **********************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'MSIAuthentication' object has no attribute 'get_token'
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible-tmp-1656865201.700019-824-22757043118455/AnsiballZ_azure_rm_virtualnetwork.py\", line 100, in <module>\n    _ansiballz_main()\n  File \"/tmp/ansible-tmp-1656865201.700019-824-22757043118455/AnsiballZ_azure_rm_virtualnetwork.py\", line 92, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/tmp/ansible-tmp-1656865201.700019-824-22757043118455/AnsiballZ_azure_rm_virtualnetwork.py\", line 41, in invoke_module\n    run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.7/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.7/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib/python3.7/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_azure_rm_virtualnetwork_payload_fplesg8l/ansible_azure_rm_virtualnetwork_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualnetwork.py\", line 386, in <module>\n  File \"/tmp/ansible_azure_rm_virtualnetwork_payload_fplesg8l/ansible_azure_rm_virtualnetwork_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualnetwork.py\", line 382, in main\n  File \"/tmp/ansible_azure_rm_virtualnetwork_payload_fplesg8l/ansible_azure_rm_virtualnetwork_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualnetwork.py\", line 240, in __init__\n  File \"/tmp/ansible_azure_rm_virtualnetwork_payload_fplesg8l/ansible_azure_rm_virtualnetwork_payload.zip/ansible_collections/azure/azcollection/plugins/module_utils/azure_rm_common.py\", line 465, in __init__\n  File \"/tmp/ansible_azure_rm_virtualnetwork_payload_fplesg8l/ansible_azure_rm_virtualnetwork_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualnetwork.py\", line 264, in exec_module\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/mgmt/network/v2021_03_01/operations/_virtual_networks_operations.py\", line 209, in get\n    pipeline_response = self._client._pipeline.run(request, stream=False, **kwargs)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/_base.py\", line 211, in run\n    return first_node.send(pipeline_request)  # type: ignore\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  [Previous line repeated 2 more times]\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/mgmt/core/policies/_base.py\", line 47, in send\n    response = self.next.send(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/policies/_redirect.py\", line 158, in send\n    response = self.next.send(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/policies/_retry.py\", line 446, in send\n    response = self.next.send(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/policies/_authentication.py\", line 117, in send\n    self.on_request(request)\n  File \"/opt/ansible/lib/python3.7/site-packages/azure/core/pipeline/policies/_authentication.py\", line 94, in on_request\n    self._token = self._credential.get_token(*self._scopes)\nAttributeError: 'MSIAuthentication' object has no attribute 'get_token'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP *****************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
`

**Comment by sharathkashyap** [comment, marked as answer]
@keshavprasadms Will you be able to help on this

**Reply by beepdot** [reply]
@harryadwani What is the ansible version being used?

**Reply by harryadwani** [reply]
This is the complete output for 'ansible --version'

`ansible [core 2.11.11] 
  config file = None
  configured module search path = ['/home/harsh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/ansible/lib/python3.7/site-packages/ansible
  ansible collection location = /home/harsh/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/ansible/bin/ansible
  python version = 3.7.3 (default, Jan 22 2021, 20:04:44) [GCC 8.3.0]
  jinja version = 3.1.2
  libyaml = True`

**Reply by harryadwani** [reply]
I have already tried this solution [https://stackoverflow.com/questions/68303285/msiauthentication-object-has-no-attribute-get-token](url), but this didn't work.

**Reply by beepdot** [reply]
Please uninstall the current ansible version and install ansible version 2.8.19 and try to run the playbook.
If you need the commands to do so, see this link - https://github.com/project-sunbird/sunbird-devops/blob/master/deploy/jenkins/jenkins-server-setup.sh#L72-L73

**Reply by harryadwani** [reply]
On azure cloud-shell, 
"pip uninstall -y ansible" gives
`WARNING: Skipping ansible as it is not installed.`

Then, running 
"pip3 install ansible==2.8.19" gives this output - 
`Defaulting to user installation because normal site-packages is not writeable
Collecting ansible==2.8.19
  Downloading ansible-2.8.19.tar.gz (12.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 18.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Requirement already satisfied: PyYAML in /usr/local/lib/python3.7/dist-packages (from ansible==2.8.19) (6.0)
Requirement already satisfied: cryptography in /usr/lib/python3/dist-packages (from ansible==2.8.19) (2.6.1)
Collecting jinja2
  Downloading Jinja2-3.1.2-py3-none-any.whl (133 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.1/133.1 kB 1.7 MB/s eta 0:00:00
Collecting MarkupSafe>=2.0
  Downloading MarkupSafe-2.1.1-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Building wheels for collected packages: ansible
  Building wheel for ansible (setup.py) ... done
  Created wheel for ansible: filename=ansible-2.8.19-py3-none-any.whl size=12656595 sha256=45733c1a3730e4b13de120fe37a1a8d2ee06a07cc4db870172644d428c9d2e67
  Stored in directory: /home/harsh/.cache/pip/wheels/87/e0/f4/530f6df144964e1c4ec7755fe715f703630ceedb7f9ee83cc8
Successfully built ansible
Installing collected packages: MarkupSafe, jinja2, ansible
Successfully installed MarkupSafe-2.1.1 ansible-2.8.19 jinja2-3.1.2`

Then, checking for ansible version again using "ansible --version" shows the previously installed ansible version.
`ansible [core 2.11.11] 
  config file = /home/harsh/sunbird-devops/deploy/ansible.cfg
  configured module search path = ['/home/harsh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/ansible/lib/python3.7/site-packages/ansible
  ansible collection location = /home/harsh/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/ansible/bin/ansible
  python version = 3.7.3 (default, Jan 22 2021, 20:04:44) [GCC 8.3.0]
  jinja version = 3.1.2
  libyaml = True`

**Reply by beepdot** [reply]
Why don't you try running the playbook from your local system after installing ansible 2.8.19 in local?

**Reply by harryadwani** [reply]
sure, I'll try that. But I'll still need to figure out a way to switch to ansible 2.8.19 in cloud-shell

**Reply by beepdot** [reply]
As cloud shell does not provide sudo / root access, you will not be able to change the default versions of packages. A workaround would be to download binaries and run them with current user path reference.

**Reply by harryadwani** [reply]
Thanks for the suggestion, I was able to change the ansible version to 2.8.19.
But I am running into a new kind of error for some reason after running `ansible-playbook -c local azure-provision.yaml
`


```

[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

ERROR! no action detected in task. This often indicates a misspelled module name, or incorrect module path.

The error appears to be in '/home/harsh/sunbird-devops/deploy/azure-provision.yaml': line 301, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

          Environment: installation
    - name: Get instance of Registry
      ^ here
```

**Reply by beepdot** [reply]
Where are you running? Are there any other outputs before this task?

**Reply by harryadwani** [reply]
Below are the only 2 commands I performed before the  `ansible-playbook -c local azure-provision.yam` after signing into my azure account and running cloud-shell. They gave their correct usual outputs.

```
git clone https://github.com/project-sunbird/sunbird-devops -b release-4.3.0
cd sunbird-devops/deploy
```

**Reply by beepdot** [reply]
The azure modules are missing in your local - https://github.com/ansible-collections/azure so none of the azure functions are being recognized.
I will give it a try to setup in a docker container the ansible and azure modules and try it out. Its been quite long since we ran this in local.
Do check this out in the mean time - https://docs.microsoft.com/en-us/azure/developer/ansible/install-on-linux-vm?tabs=azure-cli

**Reply by harryadwani** [reply]
Thanks! I resolved the error by upgrading ansible to latest then installing Azure collection hosted in Galaxy. (This was all done in cloud-shell)
However I run into another error while running `ansible-playbook -c local azure-provision.yaml`, and the error looks like I am missing a file.

```
TASK [Creating vm {{ name }}] ************************************************************************************************************************************************************************************
[WARNING]: Unable to find '/home/rajesh_r/deployer.pub' in expected paths (use -vvvvv to see paths)
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'file'. Error was a <class 'ansible.errors.AnsibleError'>, original message: could not locate file in lookup: /home/rajesh_r/deployer.pub"}

PLAY RECAP *******************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=5    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0 
```




The complete output was as follows- 

```
[DEPRECATION WARNING]: Ansible will require Python 3.8 or newer on the controller starting with Ansible 2.12. Current version: 3.7.3 (default, Jan 22 2021, 20:04:44) [GCC 8.3.0]. This feature will be removed 
from ansible-core in version 2.12. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] *************************************************************************************************************************************************************************************************

TASK [Creating resource group - "sunbird-installation-test-3.9"] *************************************************************************************************************************************************
ok: [localhost]

TASK [debug] *****************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "rg": {
        "changed": false,
        "contains_resources": false,
        "failed": false,
        "state": {
            "id": "/subscriptions/8976a67e-c189-4cc4-9cdd-4f902c5593da/resourceGroups/sunbird-installation-test-3.9",
            "location": "centralindia",
            "name": "sunbird-installation-test-3.9",
            "provisioning_state": "Succeeded",
            "tags": {}
        }
    }
}

TASK [Create virtual network] ************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Add subnet] ************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Creating public ip for jenkins] ****************************************************************************************************************************************************************************
changed: [localhost]

TASK [Creating Nics] *********************************************************************************************************************************************************************************************
changed: [localhost] => (item=jenkins)
changed: [localhost] => (item=kp)
changed: [localhost] => (item=dp)
changed: [localhost] => (item=yarn)
changed: [localhost] => (item=db)
changed: [localhost] => (item=others)
[DEPRECATION WARNING]: Setting ip_configuration flatten is deprecated and will be removed. Using ip_configurations list to define the ip configuration. This feature will be removed in version [2, 9]. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.

TASK [Assinging public ip for jenkins] ***************************************************************************************************************************************************************************
changed: [localhost]

TASK [Creating vm {{ name }}] ************************************************************************************************************************************************************************************
[WARNING]: Unable to find '/home/rajesh_r/deployer.pub' in expected paths (use -vvvvv to see paths)
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'file'. Error was a <class 'ansible.errors.AnsibleError'>, original message: could not locate file in lookup: /home/rajesh_r/deployer.pub"}

PLAY RECAP *******************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=5    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

**Reply by beepdot** [reply]
Thats great news. For this error, please change the path here to the local cloud shell path where you have the public key located
https://github.com/project-sunbird/sunbird-devops/blob/master/deploy/azure-provision.yaml#L21

**Reply by harryadwani** [reply]
Thank you so much, I created a public key from private key using ssh-keygen and was able to proceed as expected.

However, I realized that azure free trial has limited the vm cores to 4 cores for me, whereas provisioning requires 8 cores to proceed, If I understand the below error correctly.

```
TASK [Creating vm {{ name }}] ************************************************************************************************************************************************************************************
failed: [localhost] (item=jenkins) => {"ansible_loop_var": "item", "changed": false, "item": "jenkins", "msg": "Error creating or updating virtual machine jenkins - (OperationNotAllowed) Operation could not be completed as it results in exceeding approved Total Regional Cores quota. Additional details - Deployment Model: Resource Manager, Location: japaneast, Current Limit: 4, Current Usage: 4, Additional Required: 4, (Minimum) New Limit Required: 8.
```

I really wanted to try out sunbird on trial account first.

**Reply by beepdot** [reply]
The VM size is 4 cores and the playbook creates around 6 VMs each of type `Standard_D4s_v4` - https://github.com/project-sunbird/sunbird-devops/blob/master/deploy/azure-provision.yaml#L43
It looks like you have a max number of cores restriction in your account. You can talk to azure support and check if they can increase your resource quota. More reading here - https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits
In order to run Sunbird, you currently need around 6 VMs and about 4-5 nodes in Kubernetes. All together around 40-48 cores would be required for a comfortable start.
If you are just exploring sunbird, you can checkout our dev site without the need to go through the installation - https://dev.sunbird.org (some features might not work due to ongoing developement activities)
If you are planning to contribute / use in sunbird for a specific use case, do let us know.

**Reply by harryadwani** [reply]
We intend to develop a Learning Content Management System that has teacher, parent and students profile, housing diverse digital resources.

**Reply by beepdot** [reply]
Alright. As I mentioned, please check with Azure support and try to get the quota increase.

**Reply by sharathkashyap** [reply]
@harryadwani @keshavprasadms I will be closing this thread as there are multiple solutions discussed. Kindly feel free to open other discussions in case of any other queries. Locking this thread for now.
